### PR TITLE
Fix test explorer performance issues

### PR DESCRIPTION
**Pull request by kpodsiad**
fixes https://github.com/scalameta/metals/issues/3501

Previously, due to lack of caching and https://github.com/scalameta/metals/issues/3509 discovery request were able to take over 2 seconds in Metals repo.

Now, only the first request takes a significant amount of time (due to https://github.com/scalameta/metals/issues/3509 additional lookups for non-existing symbols may happen) and every subsequent request takes constant, close to zero time.

```shell
2022.01.14 14:21:18 INFO  Cache refresh took: 1752ms
2022.01.14 14:21:25 INFO  Cache refresh took: 1ms
2022.01.14 14:21:25 INFO  Cache refresh took: 1ms
```

From what I observed this [code](https://github.com/scalameta/metals/blob/5898e3b5bd4d56a4e930e37c5cb3fba8366af055/metals/src/main/scala/scala/meta/internal/metals/Compilations.scala#L220-L223) was calling refresh a few times within less than a second at the start of Metals. In order to avoid problems with unnecessary computing, I used a single-thread executor. After first run which populates caches, subsequent ones don't have much work. The executor also ensures that requests from the client will always be responded with the newest data. Maybe someone has some ideas on how to improve it?

There's one corner case that comes into my mind - if someone will rename a file without renaming the test suite then the wrong location will be cached for that test suite. I'll try to do something in this scenario.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesFinder.scala [hunk 1]
```diff
@@ -1,65 +1,137 @@
...
+
+    val time0 = System.currentTimeMillis()
+    for {
+      SymbolsPerTarget(buildTarget, testSymbols) <- symbolsPerTarget
+      (symbol, fullyQualifiedClassName) <- testSymbols.readOnlySnapshot().toList
+      // IMPORTANT this check is meant to check for class name, not a symbol
+      if !cachedTestSuites.contains(fullyQualifiedClassName)
+    } computeTestEntry(buildTarget, symbol, fullyQualifiedClassName)
+    val time1 = System.currentTimeMillis()
+    scribe.info(s"Cache refresh took: ${time1 - time0}ms")
```
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesFinder.scala:108]: Left on purpose

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesFinder.scala]: That might happen a lot, no need to leave it in since the same can be accomplished by using trace.

Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesFinder.scala]: I meant it's left on purpose for debugging reasons

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesFinder.scala]: Ok, so the plan is to remove it before the merge?

Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesFinder.scala]: Yup

File: metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+   */
+  private val cachedTestSuites =
+    mutable.Map[FullyQualifiedClassName, TestEntry]()
+
+  /**
+   * Retrieves cached test suites.
+   */
+  def findTestSuites()
+      : CompletableFuture[java.util.List[TestSuiteDiscoveryResult]] =
+    CompletableFuture.supplyAsync(
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala:55]: Is this needed? Can we instead use TrieMap and handle this normally?

File: metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    )
+
+  /**
+   * Find test suites for all build targets in current projects and update caches.
+   */
+  def refreshTestSuites(): Unit = {
+    if (clientConfig.isTestExplorerProvider()) refreshTestSuitesImpl()
+  }
+
+  def refreshTestSuitesImpl(): Unit = executor.execute { () =>
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala:80]:
```suggestion
  def refreshTestSuites(): Unit = if (clientConfig.isTestExplorerProvider()) { () =>
```
with TrieMap?

**Review by tgodzik — CHANGES_REQUESTED** [review]
Looks good, but I would not use the executor here.

File: metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala (hunk 2 above)
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala]: I used it primarily in order to avoid parallel updating of cache and it's not about synchronization, it's about duplicating work but `BatchedFunction` solves this. 
The secondary goal was to ensure that `refresh` will happen always after the end of the cache update. A situation like below could happen and test explorer won't have all of the test suites. However, I think it's not a huge issue since refresh happens very often and I already have a plan for switching approach from pull-based to push-based.
```shell
TestSuitesProvider.scala:96 "start": "start"
TestSuitesProvider.scala:56 "refresh": "refresh"
TestSuitesProvider.scala:105 "end": "end"
```

File: metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    buildTargets: BuildTargets,
+    buildTargetClasses: BuildTargetClasses,
+    definitionProvider: DefinitionProvider,
+    clientConfig: ClientConfiguration
+)(implicit ec: ExecutionContext) {
+
+  /**
+   * Cached, already discovered test suites.
+   * Cache is updated via refreshTestSuites method called after compilation is finished.
+   * There is no need to have synchronization since all access is available via
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala:36]: This is no longer true though?

**Review by tgodzik — APPROVED** [review]
LGTM! Thanks for the quick fix!

We can fix the comments later.
